doc/python_reference.rst

```diff
@@ -970,11 +971,18 @@ Simulation
    :toctree: generated/
    :template: function.rst
 
+   Simulation 
    simulate_evoked
    simulate_raw
    simulate_stc
    simulate_sparse_stc
    select_source_in_label
+   simulate_raw_signal
```

Comment:
@ngayraud can you clarify why the simulate_raw did not do the job? simulate_raw and simulate_raw_signal should be fused or one should be removed / deprecated.

Reply:
The API I have had in mind for this is that `simulate_raw` should accept `stc` as a generator instead of a single instance. Then you can pass arbitrary sets of vertices, activations, durations, etc.

Reply:
simulate raw accepts the instance of a Raw object, and I think that an Info would be better. Moreover, simulate_raw accepts an instance of SourceSpaces and Bem solution. From what I read in the code, I understood that it is important for the generatio chpi¸ blinks and ecg but isn't it too costly time-wise for it being a functions that generate simulations? simulate_raw_signal takes as input a Simulation object (new object which contains a Forward object and uses and updates its Info) and applies the forward solution  (which can be computed outside) - similar to simulate_evoked.

Reply:
> simulate raw accepts the instance of a Raw object, and I think that an Info would be better. 

If we add a `n_times` attribute or similar, this could be done. Really it just needs the `info` and number of time points to initialize the output object. So this is not a blocker.

> From what I read in the code, I understood that it is important for the generatio chpi¸ blinks and ecg but isn't it too costly time-wise for it being a functions that generate simulations

It is required when head motion is involved (head_pos=...). You cannot do this with a single forward solution. Also, a drawback of the pass-a-forward operation is that it does not guarantee that your forward is actually applicable to the `info` that you pass in (e.g., electrode locations, dev_head_t, etc.).

I also don't see this as a blocker -- we can refactor `simulate_raw` to allow passing `fwd`, and it can ensure you aren't making a mistake (ensure `src=bem=trans=head_pos=None`, and that blinks/ecg are off).

Reply:
Ok, agreed.  But maybe all the code related to computing the fwd solution (when necessary)  could be refactored into a private function - would make the code easier to read. In any case, the two functions (simulate_raw and simulate_raw_signal) are definitely going to be merged.

Reply:
Let's proceed in small steps, then. How about:

1. PR to modify make_ad_hoc_cov
2. PR to make `simulate_raw` accept `info` passed as `raw` (and in that case requiring a value for the new `n_times=None` kwarg)
3. PR to make `simulate_raw` to add a `fwd=None` kwarg
4. PR to make `simulate_raw` accept a generator for `stc`

Let me know if this plan sounds good to you. Feel free to work on these, or ping me to take one / take over if you get stuck.

Reply:
The question is: do you agree with the idea of having a Simulation object which handles everything happening before the transform between source and sensor space? In this case, the user could pass to the Simulation initialization (which could be a different function, like create_simulation or sth like that) and there he could precise all these parameters. Then, he can call any of the simulate_raw, simulate_epoch and simulate_evoke passing Simulation as one of the arguments...

Reply:
Ok thanks, I am good with the plan, I will start with (1) and let me know what you think about the rest !

Reply:
I don't think we need any more functions to the transformation from sensor->source. Thus I don't think that the forward should be necessary for it. How about step 5 as:

5. Add a `SourceSimulator` class that takes a source space and a set of vertices (or set of labels), and simulates signals for those vertices/labels with commonly used characteristics/patterns (e.g., N100-like, P300-like, etc.). It provides a `__iter__` method that returns STCs corresponding to epochs, and thus an instance can be passed to `simulate_raw(..., stc=my_source_estimator)` .

If that's the sort of thing you have in mind, then it sounds potentially quite useful. To me the target would be to have an example showing that it is very easy to simulate MNE-sample-data-like activity (L/R x aud/viz), then another that shows P300-like activity.

You mentioned `simulate_epoch`, which I don't think we really need -- this is more or less a special case of `mne.Epochs(simulate_raw(...), ...)`.

Reply:
> But then again, we would still need to compute a forward solution inside simulate_raw

No this is step (3) above